aten/src/ATen/native/cuda/ScanKernels.cu

```diff
@@ -471,14 +477,54 @@ void scan_thrust(const Tensor& self, Tensor& result, scalar_t init, BinaryFuncti
       src_data, src_data + size, dst_data,
       rocm_bug_binary_op);
   #else
-  thrust::device_ptr<scalar_t> src_data(self.data_ptr<scalar_t>());
-  thrust::device_ptr<scalar_t> dst_data(result.data_ptr<scalar_t>());
-  ptrdiff_t size = self.numel();
-  thrust::inclusive_scan(
-      thrust::cuda::par(allocator).on(c10::cuda::getCurrentCUDAStream()),
-      src_data, src_data + size, dst_data,
-      binary_op);
-  #endif
+  int64_t size = self.numel();
+  // non synchronizing cub call
+  // even though cub is supposed to support tensors with int_max elements, in reality it doesn't,
+  // so split at int_max/2
+  constexpr int max_cub_size = std::numeric_limits<int>::max() / 2 + 1; // 2**30
+  for (int64_t i = 0; i < size; i += max_cub_size) {
+    int size_cub = std::min<int64_t>(size - i, max_cub_size);
+    Tensor first_elem; // need to save it for all iterations other than first
+    if (i > 0) {
+      // need to temporarily transform first element of the range we are
+      // operating on self might be multi-d, but we need to index a single
+      // element
+      auto self_view = at::_unsafe_view(self, -1);
```

Comment:
Why view self inside the loop?

Reply:
In most cases we won't enter this loop, in some others we'll enter it once. Taking the view out of the loop will make us take it unconditionally, which we don't want.